extensions/ql-vscode/src/remote-queries/view/AnalysisAlertResult.tsx

```diff
@@ -0,0 +1,171 @@
+const warningColor = '#966C23';
+const highlightColor = '#534425';
+
+const getSeverityColor = (severity: ResultSeverity) => {
+  switch (severity) {
+    case 'Recommendation':
+      return 'blue';
+    case 'Warning':
+      return warningColor;
+    case 'Error':
+      return 'red';
+  }
+};
+
+const Container = styled.div`
+  font-family: ui-monospace, SFMono-Regular, SF Mono, Menlo, Consolas, Liberation Mono, monospace;
+`;
+
+const TitleContainer = styled.div`
+  border: 0.1em solid ${borderColor};
+  border-top-left-radius: 0.2em;
+  border-top-right-radius: 0.2em;
+  padding: 0.5em;
+`;
+
+const CodeContainer = styled.div`
+  font-size: x-small;
+  border-left: 0.1em solid ${borderColor};
+  border-right: 0.1em solid ${borderColor};
+  border-bottom: 0.1em solid ${borderColor};
+  border-bottom-left-radius: 0.2em;
+  border-bottom-right-radius: 0.2em;
+  padding-top: 1em;
+  padding-bottom: 1em;
+`;
+
+const MessageText = styled.span<{ severity: ResultSeverity }>`
+  font-size: x-small;
+  color: ${props => getSeverityColor(props.severity)};
+  padding-left: 0.5em;
+`;
+
+const MessageContainer = styled.div`
+  padding-top: 0.5em;
+  padding-bottom: 0.5em;
+`;
+
+const Message = ({ alert, currentLineNumber }: {
+  alert: AnalysisAlert,
+  currentLineNumber: number
+}) => {
+  if (alert.highlightedRegion.startLine !== currentLineNumber) {
+    return <></>;
+  }
+  return <MessageContainer>
+    <Box
+      borderColor="border.default"
+      borderWidth={1}
+      borderStyle="solid"
+      borderLeftColor={getSeverityColor(alert.severity)}
+      borderLeftWidth={3}
+      paddingTop="1em"
+      paddingBottom="1em">
+      <MessageText severity={alert.severity}>{alert.message}</MessageText>
+    </Box>
+
+  </MessageContainer>;
+};
+
+const replaceSpaceChar = (text: string) => text.replaceAll(' ', '\u00a0');
+
+const PlainLine = ({ text }: { text: string }) => {
+  return <span>{replaceSpaceChar(text)}</span>;
+};
+
+const HighlightedLine = ({ text }: { text: string }) => {
+  return <span style={{ backgroundColor: highlightColor }}>{replaceSpaceChar(text)}</span>;
+};
+
+const shouldHighlightLine = (lineNumber: number, highlightedRegion: HighlightedRegion) => {
+  if (lineNumber < highlightedRegion.startLine) {
+    return false;
+  }
+
+  if (highlightedRegion.endLine) {
+    return lineNumber <= highlightedRegion.endLine;
+  }
+
+  return true;
+};
+
+const CodeLine = ({
+  line,
+  lineNumber,
+  highlightedRegion
+}: {
+  line: string,
+  lineNumber: number,
+  highlightedRegion: HighlightedRegion
+}) => {
+  if (!shouldHighlightLine(lineNumber, highlightedRegion)) {
+    return <PlainLine text={line} />;
+  }
+
+  const section1 = line.substring(0, highlightedRegion.startColumn - 1);
+  const section2 = line.substring(highlightedRegion.startColumn - 1, highlightedRegion.endColumn - 1);
+  const section3 = line.substring(highlightedRegion.endColumn - 1, line.length);
```

Comment:
Make sure that you can gracefully handle invalid regions. These regions can be created by user queries which could return invalid source locations. Eg- `endLine < startLine` or `endLine == startLine && endColumn <= startColumn`. In general (I believe) that end column must be strictly greater than start column on the same line.

Also, note that CodeQL/Sarif source locations have a slightly different semantics than vscode locations. See here for how it is handled for local queries: https://github.com/github/vscode-codeql/blob/735f177283ed7e02674ed6ff1af889d61f9a878c/extensions/ql-vscode/src/interface-utils.ts#L64-L78

I don't know if this will affect remote queries since everything is self contained, but just something to be aware of.

It would be fine to just keep the handling simpler for now and make it more robust in a future PR.

Reply:
Thanks! I've added some basic validation but will probably revisit to tidy things up in later PR.